src/Voron/Data/CompactTrees/PersistentDictionaryLocator.cs

```diff
@@ -65,14 +55,33 @@ public void Set(long pageNumber, PersistentDictionary dictionary)
         {
             var position = pageNumber & _andMask;
 
+            // This is intentionally racy because we are trying to optimize the TryGet method
+            // as much as possible, we are fine with "losing" cached instances, they will make themselves
+            // up over time and fix themselves
+
             // No check need to be done at this level because at the read side we will
-            // make sure to get the proper dictionary. It doesn't matter in which order
-            // we store the page and the dictionary since race conditions are guaranteed
-            // to happen at this level. That's why we do a double check on the dictionary
-            // we retrieve itself on the TryGet method.
-            ref var item = ref _cache[position];
-            item.Dictionary = dictionary;
-            item.PageNumber = pageNumber;
+            // make sure to get the proper dictionary. 
+            switch (_cache[position])
+            {
+                case null:
+                    _cache[position] = dictionary;
+                    break;
+                case PersistentDictionary existing:
+                    if (existing.DictionaryId == dictionary.DictionaryId)
+                        return; // should never happen
+                    _cache[position] = new Dictionary<long, PersistentDictionary>
+                    {
+                        [existing.DictionaryId] = existing,
+                        [dictionary.DictionaryId] = dictionary
```

Comment:
Subtle behavior here. 
If we have two dictionaries that are on the _same_ id, we'll use the `dictionary` one, not the `existing`, it is not an error in this case.